src/typeahead/typeahead.spec.ts

```diff
@@ -194,6 +197,93 @@ describe('ngb-typeahead', () => {
       expect(getWindow(compiled)).not.toBeNull();
     });
 
+    it('should open on focus or click', async(async() => {
+         const fixture = createTestComponent(`<input
+        type="text"
+        [ngbTypeahead]="find"
+        (focus)="focus$.next($event.target.value)"
+        (click)="click$.next($event.target.value)"
+      />`);
+         const compiled = fixture.nativeElement;
+
+         let searchCount = 0;
+         fixture.componentInstance.findOutput$.subscribe(() => searchCount++);
+         const checkSearchCount = (expected, context) => {
+           expect(searchCount).toBe(expected, `Search count is not correct: ${context}`);
+           searchCount = 0;
+         };
+
+         const checkWindowIsClosed = () => {
+           expect(getWindow(compiled)).toBeNull();
+           expect(fixture.componentInstance.model).toBe(undefined);
+           expect(getNativeInput(compiled).value).toBe('');
+         };
+
+         const checkWindowIsOpen = () => { expect(getWindow(compiled)).not.toBeNull(); };
+
+         // focusing the input triggers a search and opens the dropdown
+         getNativeInput(compiled).focus();
+         // on IE the focus can be asynchronous, so we need to wait a tick before continuing
+         await new Promise(resolve => setTimeout(resolve, 0));
```

Comment:
Wouldn't `Promise.resolve` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve) work here?

More importantly, do we need to trigger native `focus` event here? Wouldn't Angular testing trigger event be enough?

Reply:
Sure, initially I tried with a timeout higher than 0, so I didn't realize that for this "next tick" use case using `Promise.resolve()` would work.

Regarding the native `focus` event it's indeed because we don't use a regular Angular event handler, it work with Observables on the user side.